lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -604,46 +609,106 @@ void OCLBackend::doForwardPass() {
     clReleaseKernel(kl.kernel_);
   }
 
-  copyWeightsFromDevice();
+  auto copiedFromDeviceBytes = copyMutableWeightsFromDevice();
+  (void)copiedFromDeviceBytes;
+  DEBUG(llvm::dbgs() << "Copied " << copiedFromDeviceBytes
+                     << " bytes from OpenCL device\n");
 }
 
-void OCLBackend::copyWeightsToDevice() {
+size_t OCLBackend::copyValueToDevice(const Value *v, void *buf) {
+  size_t copiedBytes = 0;
+  auto it = tensors_.find(v);
+  assert(it != tensors_.end() && "Unknown value");
+  size_t sizeInBytes = v->getType()->getSizeInBytes();
+  // Issue a non-blocking command to copy the buffer to the device.
+  if (sizeInBytes) {
+    if (!buf) {
+      Tensor *T = externalTensors_[v];
+      assert(T && "Expectded an external tensor");
+      buf = T->getUnsafePtr();
+    }
+    cl_int err =
+        clEnqueueWriteBuffer(commands_, deviceBuffer_, CL_FALSE, it->second,
+                             sizeInBytes, buf, 0, nullptr, nullptr);
+    GLOW_ASSERT(err == CL_SUCCESS && "Unable to copy data to the device");
+    copiedBytes += sizeInBytes;
+  }
+  return copiedBytes;
+}
+
+size_t OCLBackend::copyValueFromDevice(const Value *v, void *buf) {
+  size_t copiedBytes = 0;
+  auto it = tensors_.find(v);
+  assert(it != tensors_.end() && "Unknown value");
+  size_t sizeInBytes = v->getType()->getSizeInBytes();
+  // Issue a non-blocking command to copy the buffer from the device.
+  if (sizeInBytes) {
+    if (!buf) {
+      Tensor *T = externalTensors_[v];
+      assert(T && "Expectded an external tensor");
+      buf = T->getUnsafePtr();
+    }
+    cl_int err =
+        clEnqueueReadBuffer(commands_, deviceBuffer_, CL_FALSE, it->second,
+                            sizeInBytes, buf, 0, nullptr, nullptr);
```

Comment:
For the constant params here, perhaps add some comment about what they are/outline these params with useful variable names?